src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlRowValueComparisonTranslator.cs

```diff
@@ -0,0 +1,93 @@
+            .Single(m => m.Name == nameof(NpgsqlDbFunctionsExtensions.RowValueLessThanOrEqual));
+
+    private static readonly Dictionary<MethodInfo, ExpressionType> MethodDefinitions = new()
+    {
+        { GreaterThan, ExpressionType.GreaterThan },
+        { LessThan, ExpressionType.LessThan },
+        { GreaterThanOrEqual, ExpressionType.GreaterThanOrEqual },
+        { LessThanOrEqual, ExpressionType.LessThanOrEqual }
+    };
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="NpgsqlRowValueComparisonTranslator"/> class.
+    /// </summary>
+    public NpgsqlRowValueComparisonTranslator(NpgsqlSqlExpressionFactory sqlExpressionFactory)
+        => _sqlExpressionFactory = sqlExpressionFactory;
+
+    /// <inheritdoc />
+    public virtual SqlExpression? Translate(
+        SqlExpression? instance,
+        MethodInfo method,
+        IReadOnlyList<SqlExpression> arguments,
+        IDiagnosticsLogger<DbLoggerCategory.Query> logger)
+    {
+        if (method.DeclaringType != typeof(NpgsqlDbFunctionsExtensions)
+            || !method.IsGenericMethod
+            || !MethodDefinitions.TryGetValue(method.GetGenericMethodDefinition(), out var expressionType))
+        {
+            return null;
+        }
+
+        var (rowValue1, rowValue2) = (GetRowValue(arguments[1]), GetRowValue(arguments[2]));
+
+        if (rowValue1.Values.Count != rowValue2.Values.Count)
+        {
+            throw new ArgumentException(NpgsqlStrings.RowValueMethodRequiresTwoArraysOfSameLength(method.Name));
+        }
+
+        return _sqlExpressionFactory.MakeBinary(expressionType, rowValue1, rowValue2, typeMapping: null);
+
+        PostgresRowValueExpression GetRowValue(SqlExpression arrayExpression)
+        {
+            switch (arrayExpression)
+            {
+                case PostgresNewArrayExpression newArray:
+                    // Unwrap any Convert-to-object nodes
+                    var unwrappedExpressions = new SqlExpression[newArray.Expressions.Count];
+
+                    for (var i = 0; i < unwrappedExpressions.Length; i++)
+                    {
+                        unwrappedExpressions[i] =
+                            newArray.Expressions[i] is SqlUnaryExpression { OperatorType: ExpressionType.Convert } convertExpression
+                            && convertExpression.Type == typeof(object)
+                                ? convertExpression.Operand
+                                : newArray.Expressions[i];
+                    }
+
+                    return new PostgresRowValueExpression(unwrappedExpressions);
+
+                case SqlConstantExpression { Value : object[] constant }:
+                    return new PostgresRowValueExpression(constant.Select(c => _sqlExpressionFactory.Constant(c)).ToArray());
+
+                default:
+                    throw new ArgumentException($"{method.Name} accepts only new array initialization literals");
```

Comment:
does this mean parameters are not allowed?

Reply:
Unfortunately it does, at the moment... Apparently a LINQ NewArrayExpression gets parameterized the moment it contains a parameter inside it (if we somehow got the NewArrayExpression with ParameterExpression inside it, that would work).

As for handling a ParameterExpression representing the entire row value... PG does have "composite types" (like UDT, which represents a row in a table). At the ADO level Npgsql supports mapping these to CLR types and reading/writing them, but there's no composite support at the EF provider level (#22). If it's valuable/requested I can always add that later.

Note that composite parameters would also be PG-specific: while SQLite and MySQL support row values "literals" (as supported via this PR), I don't think they have a true 1st-class type like that, so a parameterized row value is probably impossible.

Makes sense?

Reply:
It's true this restriction does limit the usefulness of this feature, since typically you'd want to parameterize the constants on the right-hand side...

Reply:
So yeah... PostgreSQL doesn't currently allow parameterizing "anonymous" row types (or records), as in https://github.com/npgsql/npgsql/issues/4437#issuecomment-1114029098. Named row types (AKA composites) can be parameterized, but that means a type must exist prior to the query, which is inappropriate here (similar to how we can't use SQL Server TVPs in various scenarios).

I can add `[NotParameterizable]` to the array parameters of these new functions. This stops us from trying to parameterize the entire array, but also seems to prevent parameterization of the elements inside the array, making them into constants instead. So:

```c#
var foo = "foo";
_ = ctx.Blogs.Where(b => EF.Functions.RowValueGreaterThan(new[] { b.Name1, b.Name2 }, new[] { "A", foo })).ToList();
```

... yields:

```sql
SELECT b."Id", b."Name1", b."Name2"
FROM "Blogs" AS b
WHERE (b."Name1", b."Name2") > ('A', 'foo')
```

(where 'foo' should be a parameter instead of a constant).

Reply:
Just excluding all NewArrayExpression in EvaluatableExpressionFilter produces the right results (`('A', @__foo_1)`), but seems like too much... In theory if we had something like [NotParameterizable] but with more options, e.g. to prevent evaluability on a specific parameter...

OK enough, going to bed... :sleeping:

Reply:
Sounds fine to me, just wanted to make sure you have considered parameters also. Can always add it later if there is user feedback and an easy way.

Reply:
Yeah, looking at this again with fresh eyes, I guess excluding NewArrayExpression via EvaluatableExpressionFilter isn't too bad (I'm assuming evaluatable NewArrayExpressions should be rare?). Otherwise, if parameters aren't supported, this feature isn't very useful...

If you think otherwise or have some other bright idea, let me know.